File: 276. Paint Fence.py
class Solution:
    def numWays(self, n: int, k: int) -> int:
        '''
        Thought is like followings:
        1. if prev is as same color as prev of prev:
            then we cannot choose that color (k-1 choice)
        2. if prev is a diff color as prev of prev:
            then we:
                1. choose the same as prev
                2. choose diff color
        
        So we keep two vars:
            same: number of prev is as prev of prev
            diff: number of prev is diff as prev of prev

        Update rule:
            same = prev_diff
            diff = prev_same * (k-1) + prev_diff * (k-1)

        Time O(n)
        Space O(1)
        '''
        if n == 0:
            return 0
        if n == 1:
            return k
        same = 0
        diff = k
        p = 1
        while p < n:
            same, diff = diff * 1, diff * (k-1) + same * (k-1)
            p += 1
        return same+diff

File: 276. Paint Fence.java
/**
 * dp solution, in each step:
 * (1) if we want to use the same color as prev one, we can only use the number of prev diff, curr_same = prev_diff
 * (2) if we want to use the diff color as prev one, we can use diff and the same and choose the different color as prev, 
 *     curr_diff = prev_same*(k-1) + prev_diff*(k-1)
 */
class Solution {
    public int numWays(int n, int k) {
        if (n * k==0) {
            return 0;
        }
        int same = 0;
        int diff = k;
        for (int i=1; i<n; i++) {
            int sametemp = diff;
            int difftemp = same * (k-1) + diff * (k-1);
            same = sametemp;
            diff = difftemp;
        }
        return same + diff;
    }
}